packages/lesswrong/components/posts/PostsNewForm.tsx

```diff
@@ -321,7 +369,11 @@ const PostsNewForm = ({classes}: {
                 successCallback={(post: any, options: any) => {
                   if (!post.draft) afNonMemberSuccessHandling({currentUser, document: post, openDialog, updateDocument: updatePost});
                   if (options?.submitOptions?.noReload) {
-                    navigate(postGetEditUrl(post._id, true), { replace: true });
+                    // First prefetch the relevant post fragments to hydrate the apollo cache, then do the navigation after that's done
+                    void prefetchPostFragmentsForRedirect(post._id).then(() => {
+                      const editPostUrl = `${postGetEditUrl(post._id, false, post.linkSharingKey)}&autosaveRedirect=true`;
+                      navigate(editPostUrl, { replace: true });
+                    });
```

Comment:
This is the only meaningful change inside of the component itself (as well as the new hook); the rest was me moving things around because the previous state of affairs was making me very sad.